### PR TITLE
feat(cli): Webhook listener

### PR DESCRIPTION
**Pull request by morrieinmaas**
wip

This works but needs some tlc

Signed-off-by: Moriarty <moritz@animo.id>

### CODE REVIEW

**Comment by morrieinmaas** [comment]
closes #214

**Comment by morrieinmaas** [comment]
Probably should add this to the docs as well @blu3beri - will do when no other changes required.

File: Cargo.toml [hunk 1]
```diff
@@ -30,6 +30,7 @@ automations = { path = "./crates/automations" }
 async-trait = "0.1.51"
 serde = { version = "1.0.130", features = ["derive"] }
 logger = { path = "./crates/logger" }
+webhooks = { path = "./crates/webhooks" }
```
Comment (blu3beri) [inline Cargo.toml:33]: Don't know how I feel about it being a separate crate. Does not really make sense to me.

File: crates/cli/src/help_strings.rs [hunk 2]
```diff
@@ -145,6 +149,8 @@ impl HelpStrings {
             Self::Config => "Supply a path to your configuration file to use that instead of the default",
             Self::Environment => "Specify your current environment",
 
+            Self::Webhooks => "Listen to webhooks",
+            Self::WebhooksEnvironment => "The environment from the config to use for listening to webhooks",
```
Comment (blu3beri) [inline crates/cli/src/help_strings.rs:153]: This should not be added here. the core `cli` package manages state here. It should just be passed in like

```
agent-cli --environment="foo" listen
```

and not

```
agent-cli listen --environment="foo"
```

File: crates/cli/src/modules/webhooks.rs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    },
+}
+
+/// Subcommand Webhooks parser
+pub async fn parse_webhooks_args(
+    options: &WebhooksOptions,
+) -> Result<()> {
+    let loader = Loader::start(&LoaderVariant::default());
+
+    match &options.commands {
```
Comment (blu3beri) [inline crates/cli/src/modules/webhooks.rs:38]: This should not be done here and is done by the base `agent-cli` and not command specific.

File: crates/webhooks/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    
+    let (mut socket, _response) = connect(Url::parse(&listen_url).unwrap()).expect("Can't connect to websocket");    // Write a message containing "Hello, Test!" to the server
+    
+    // Loop forever, parse message to stdout
+    loop {
+        // TODO: Replace with error string from enum
+        let msg = socket.read_message().expect("Error reading message");
+        let msg = match msg {
+            Message::Text(s) => { s }
+            _ => { panic!() }
```
Comment (blu3beri) [inline crates/webhooks/src/lib.rs:26]: `unreachable!()` is preferred here.

File: crates/webhooks/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    let stripped_agent_url = match &agent_url {
+        s if s.starts_with("http://") => &s[7..], 
+        s if s.starts_with("https://") => &s[8..],
+        s => s,
+    };
+
+    let listen_url = format!("wss://{}/ws", stripped_agent_url);
+    println!("Listening on {}\n", listen_url);
+    
+    let (mut socket, _response) = connect(Url::parse(&listen_url).unwrap()).expect("Can't connect to websocket");    // Write a message containing "Hello, Test!" to the server
```
Comment (blu3beri) [inline crates/webhooks/src/lib.rs:18]: All the expects should be removed and transformed into returning in `Result`

File: crates/webhooks/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+pub fn listen(agent_url: String) -> ! {
+    // TODO: filter by/listen to by topic
+    let stripped_agent_url = match &agent_url {
+        s if s.starts_with("http://") => &s[7..], 
+        s if s.starts_with("https://") => &s[8..],
+        s => s,
+    };
+
+    let listen_url = format!("wss://{}/ws", stripped_agent_url);
+    println!("Listening on {}\n", listen_url);
```
Comment (blu3beri) [inline crates/webhooks/src/lib.rs:16]: Should use the logger crate.

File: crates/webhooks/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    loop {
+        // TODO: Replace with error string from enum
+        let msg = socket.read_message().expect("Error reading message");
+        let msg = match msg {
+            Message::Text(s) => { s }
+            _ => { panic!() }
+        };
+        // TODO: Replace with error string from enum
+        let parsed: Value = from_str(&msg).expect("Can't parse to JSON");
+        println!("{}{}", ("Received hook:\n").yellow(), to_string_pretty(&parsed).unwrap());
```
Comment (blu3beri) [inline crates/webhooks/src/lib.rs:30]: Should use print.

**Review by blu3beri — CHANGES_REQUESTED** [review]
I like the addition, very useful and @karimStekelenburg will be happy! However the implementation needs some changes. 

1. Webhooks should not be a crate
2. Environment should be retrieved from the webhookcommand but from the agent-cli.
3. The `listen` function should use the logger crate for printing.
4. The `listen` function MUST not panic. 
5. `listen` function needs some refactoring but that can be done in a separate PR.

File: Cargo.toml (hunk 1 above)
Comment (morrieinmaas) [reply Cargo.toml:33]: Where else would you put it?

Comment (morrieinmaas) [reply Cargo.toml:33]: just as a module in cli i guess?!

File: crates/webhooks/src/lib.rs (hunk 7 above)
Comment (morrieinmaas) [reply crates/webhooks/src/lib.rs:30]: could also use the logger here @blu3beri ?

File: Cargo.toml (hunk 1 above)
Comment (blu3beri) [reply Cargo.toml:33]: yes.

File: crates/cli/src/modules/webhooks.rs [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+#[derive(Subcommand, Debug)]
+#[clap(about = HelpStrings::Webhooks)]
+pub enum WebhooksSubcommands {
+    /// Listen for webhooks on provided url
+    #[clap(about = HelpStrings::Webhooks)]
+    Listen {},
+}
+
+/// Listen to webhooks for an agentURL
+pub fn listen(agent_url: String) -> ! {
```
Comment (blu3beri) [inline crates/cli/src/modules/webhooks.rs:29]: I am not a fan of the `!` return in this case.

As I mentioned before, we should return a result so we can handle errors like we are doing in every submodule.

```suggestion
pub fn listen(agent_url: String) -> Result<()> {
```

This makes it very easy to handle errors in a consistent way across all the modules.

Using `unreachable!` and `unwrap` is extremely discouraged and there must be a very good reason for it to be in the codebase.


In this case we should just bubble-up the error and deal it in a general way.

File: crates/cli/src/modules/webhooks.rs [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    // TODO: filter by/listen to by topic
+    let stripped_agent_url = match &agent_url {
+        s if s.starts_with("http://") => &s[7..], 
+        s if s.starts_with("https://") => &s[8..],
+        s => s,
```
Comment (blu3beri) [inline crates/cli/src/modules/webhooks.rs:34]: This is not correct and will panic below.

Is neither of the cases are met we should return an error and bubble it up.

File: crates/cli/src/modules/webhooks.rs [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    };
+
+    let listen_url = format!("wss://{}/ws", stripped_agent_url);
+    log!("Listening on {}\n", listen_url);
+    
+    let (mut socket, _response) = match connect(Url::parse(&listen_url).unwrap()) {
```
Comment (blu3beri) [inline crates/cli/src/modules/webhooks.rs:40]:
```suggestion
    let (mut socket, _response) = match connect(Url::parse(&listen_url)?) {
```

Likely needs an error conversion to a custom agent-cli error.

File: crates/cli/src/modules/webhooks.rs [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+        Ok((socket, _response)) => (socket, _response),
+        Err(e) => {log!("{}", e); unreachable!()}
```
Comment (blu3beri) [inline crates/cli/src/modules/webhooks.rs:42]: This should not be done here and the `connect` function just needs another error conversion and be called like `connect(...)?`.

File: crates/cli/src/modules/webhooks.rs [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    };
+    
+    // Loop forever, parse message to stdout
+    loop {
+        // TODO: Replace with error string from enum
+        let msg = match socket.read_message() {
```
Comment (blu3beri) [inline crates/cli/src/modules/webhooks.rs:48]: Same here. No error case and it should be like this:

```suggestion
        let msg = match socket.read_message()?;
```

File: crates/cli/src/modules/webhooks.rs [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+            Ok(m) => m,
+            Err(e) => {log!("{}", e); unreachable!()}
+        };
+        let msg = match msg {
+            Message::Text(s) => { s }
+            _ => { unreachable!() }
+        };
+        // TODO: Replace with error string from enum
+        let parsed: Value = from_str(&msg).expect("Can't parse to JSON");
```
Comment (blu3beri) [inline crates/cli/src/modules/webhooks.rs:57]: no expect.

File: crates/cli/src/modules/webhooks.rs [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+        log!("{}{}\n", ("Received hook:\n").yellow(), to_string_pretty(&parsed).unwrap());
```
Comment (blu3beri) [inline crates/cli/src/modules/webhooks.rs:58]: no unwrap.

and a nicer way would be two logs:

```rust
log!("Received hook:\n");
log!("{}", ...);
```

File: crates/cli/src/register.rs [hunk 15]
```diff
@@ -64,6 +65,7 @@ pub async fn register() -> Result<()> {
                 match &cli.commands {
                     Commands::Schema(options) => parse_schema_args(options, agent).await,
                     Commands::Feature(_) => parse_features_args(agent).await,
+                    Commands::Webhooks(_) => {listen(agent.endpoint)},
```
Comment (blu3beri) [inline crates/cli/src/register.rs:68]: for convince in patterns just pass a reference to the entire agent. Seems a bit odd to do it differently here.

File: crates/cli/src/modules/webhooks.rs [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+    };
+    
+    // Loop forever, parse message to stdout
+    loop {
+        // TODO: Replace with error string from enum
+        let msg = match socket.read_message() {
+            Ok(m) => m,
+            Err(e) => {log!("{}", e); unreachable!()}
+        };
+        let msg = match msg {
```
Comment (blu3beri) [inline crates/cli/src/modules/webhooks.rs:52]: Same here. `unreachable!`, my fault for recommending it, is an undesirable pattern and should be avoided in cases where we can (like here).

**Review by blu3beri — CHANGES_REQUESTED** [review]
Good changes, but the `listen` function is not consistent with the other modules or inline with idiomatic Rust.

**Comment by blu3beri** [comment]
@morrieinmaas If you have no time for this I can also apply the fixes myself.

**Comment by morrieinmaas** [comment]
Thx for all the feedback @blu3beri ! If it's not urgent Id rather do it myself to get some practice in. If it is urgent feel free to go ahead

File: crates/cli/src/modules/webhooks.rs (hunk 9 above)
Comment (morrieinmaas) [reply crates/cli/src/modules/webhooks.rs:34]: hmm I see. Was thinking that there might be a url without prefix in there... But I guess you do have a point here - better to assume the url is wrong altogether

File: crates/cli/src/modules/webhooks.rs (hunk 8 above)
Comment (morrieinmaas) [reply crates/cli/src/modules/webhooks.rs:29]: I do get your point - however, wondering whether Result is the right return type cos it just prints to void until ctrl-c so it's more like void or error?! Dunno (*newbie facepalm*)

Comment (morrieinmaas) [reply crates/cli/src/modules/webhooks.rs:29]: so not sure how Ok is returned

Comment (blu3beri) [reply crates/cli/src/modules/webhooks.rs:29]: Ok(()) should be the last line of your code and will be returned when the loop exits for any reason. `<C-c>` will short-circuit. 

Your `void or error` is exactly what a `Result<()>` means.

return nil, or void on Ok
return Error on Error

**Comment by morrieinmaas** [comment]
Alright did my best to finish it but there is the return type issue left. Feel free to do the rest and merge if you want as I'm out until Tuesday.

Cheers

**Review by blu3beri — APPROVED** [review]

**Pull request merged**